### ISSUE TITLE
[Bug][PTOAS][a5] `pto.tcmp` verifier demands src/dst same shape, which no packed predicate mask can satisfy — blocks 3 mtp cases on a5sim

### ISSUE THREAD

**Issue by Crystal-wzy** [zhangstevenunity/PTOAS#1450]
### Diagnosis

**PTOAS — the a5 `pto.tcmp` operation verifier** (not pto-isa, not pypto).

Three `deepseek_v4_flash_mtp` cases fail on `a5sim` during code generation:

```text
Failed to compile function 'sample_gumbel_argmax': ptoas compilation failed:
loc(".../models/deepseek_v4_flash_mtp/sample.py":172:9):
error: 'pto.tcmp' op expects src0, src1, and dst to have the same shape
```

`sample.py:172` is a plain tile-vs-tile compare inside `_greedy_sample_logits`:

```python
scores = logits_grid[block_row : block_row + GREEDY_BLOCK_ROWS_TILE, 0:GREEDY_ROW_WIDTH_TILE]
is_newer = pl.cmp(scores, running_max, cmp_type=4)     # line 172
```

with `GREEDY_BLOCK_ROWS_TILE = 8`, `GREEDY_ROW_WIDTH_TILE = 808`, so both operands are `[8, 808]` FP32.

**The emitted IR is well-formed, and identical on both targets.** pypto lowers it to a packed predicate mask:

```mlir
pto.tcmp ins(%scores, %running_max {cmpMode = #pto<cmp gt>}
             : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=808, ...>,
               !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=808, ...>)
         outs(%cmp_mask : !pto.tile_buf<loc=vec, dtype=ui8, rows=8, cols=128, ...>)
```

`cols=128` is the mask packing: 808 predicate bits = 101 bytes, rounded up to the 32-byte block granularity
= 128. Compiling the same source for `a2a3sim` emits a **byte-identical** `outs(...)` type — and succeeds.

**Three independent facts show the verifier is the defect:**

1. **Same IR, different verdict by target.** `a2a3sim` produces `ptoas/sample_gumbel_argmax.cpp` (PTOAS
   accepted it); `a5sim` produces only `ptoas/sample_gumbel_argmax.pto` and stops. Nothing in the IR differs.
2. **a5 already accepts this mask convention for the sibling op.** In the *same source file and the same a5
   build*, `pto.tcmps` (tile-vs-scalar compare) passes with exactly the same packing —
   `[8, 640] f32 → [8, 96] ui8` (640 bits = 80 bytes → 96 after alignment) — and
   `ptoas/sample_apply_top_k.cpp` is emitted. Only `tcmp` is rejected.
3. **The constraint is unsatisfiable by construction.** A packed mask can never have the source's shape:
   for `cols=808`, the mask is 101 bytes → 128 after 32-byte alignment, or 32 `u32` words. Neither is 808.
   No legal packed predicate mask, of any dtype, can make `dst` shape-equal to `src0`. The check as worded
   cannot be satisfied by a conforming program.

**pto-isa specifies the packed, differently-shaped mask** (`docs/isa/TCMP.md`):

> Compare two tiles and write a packed predicate mask. […] The predicate mask is stored in `dst` using an
> implementation-defined packed layout.

and its own canonical example ships a `dst` whose declared shape differs from `src`:

```cpp
using SrcT  = Tile<TileType::Vec, float,   16, 16>;
using MaskT = Tile<TileType::Vec, uint8_t, 16, 32, BLayout::RowMajor, -1, -1>;
SrcT src0, src1;
MaskT mask(16, 2);
TCMP(mask, src0, src1, CmpMode::GT);
```

The documented equality is on the *valid* extents (`src0.GetValidRow() == dst.GetValidRow()` and the same for
cols), not on the declared tile shape that the MLIR verifier is checking.

### Source

- Parent CI run: https://github.com/hw-native-sys/pypto-lib/actions/runs/33453529066
- Workflow: `Daily CI #187` (scheduled 2026-09-01 00:08 UTC)
- PyPTO-Lib branch / commit: `main` / `0e7e07aa5e42db6e08c68fd5bfe1def571492b53`
- Job: `model-tests-sim (a5sim)` — 3 affected cases
- Reproduced locally on pto-isa `master` `39782974cea3b8068b428b43df8f0aaceb0925e4`

All three reach one function, `sample_gumbel_argmax` in `models/deepseek_v4_flash_mtp/sample.py`:

| case | path to the failure | a5sim | a2a3sim |
|---|---|---|---|
| `mtp/sample.py` | directly | `pto.tcmp` shape error | PASS |
| `mtp/lm_head.py` | `from sample import golden_sample, sample` | same | PASS |
| `mtp/decode_mtp.py` | `from lm_head import ...` -> `sample` | same | PASS |

### Failure Details From Logs

```text
1 function(s) failed to compile:

  Function              | Error
  ----------------------+--------------------------------------------------------------
  sample_gumbel_argmax  | ptoas compilation failed: loc(".../sample.py":172:9):
                        | error: 'pto.tcmp' op expects src0, src1, and dst to have the
                        | same shape
Error: Failed to parse MLIR.
```

Per-function outcome in one a5 build directory — the contrast that isolates the op:

| generated function | contains | a5 result |
|---|---|---|
| `sample_apply_temperature` | no compare | `.cpp` emitted |
| `sample_apply_top_k` | `pto.tcmps`, mask `[8, 640] f32 -> [8, 96] ui8` | `.cpp` emitted |
| `sample_gumbel_argmax` | `pto.tcmp`, mask `[8, 808] f32 -> [8, 128] ui8` | **`.pto` only — rejected** |

### Reproduction

**Minimal — PTOAS alone, no pypto, no model.** Two f32 tiles and the packed mask they compare into:

```mlir
// packed_mask.pto
module {
  func.func private @tcmp_packed_mask(
      %src0: !pto.tile_buf<vec, 8x808xf32>,
      %src1: !pto.tile_buf<vec, 8x808xf32>,
      %dst: !pto.tile_buf<vec, 8x128xi8>) {
    pto.tcmp ins(%src0, %src1 {cmpMode = #pto<cmp gt>} : !pto.tile_buf<vec, 8x808xf32>, !pto.tile_buf<vec, 8x808xf32>)
             outs(%dst : !pto.tile_buf<vec, 8x128xi8>)
    return
  }
}
```

```bash
ptoas --pto-level=level3 --pto-arch=a5 packed_mask.pto -o /dev/null
# -> error: 'pto.tcmp' op expects src0, src1, and dst to have the same shape
```

Swap `pto.tcmp` for `pto.tcmps` (same tiles, a scalar operand) and it compiles — that one-token edit isolates
the defect to the `tcmp` verifier.

**End-to-end:**

```bash
python models/deepseek_v4_flash_mtp/sample.py -p a5sim  --compile-only
# -> 'pto.tcmp' op expects src0, src1, and dst to have the same shape

python models/deepseek_v4_flash_mtp/sample.py -p a2a3sim --compile-only
# -> [RUN] PASS (0.54s)
```

Compare the emitted `outs(...)` type in
`build_output/_jit_sample_test_*/ptoas/sample_gumbel_argmax.pto` from the two runs: they are identical, so
the divergence is entirely in the verifier.

Any tile-vs-tile `pl.cmp` reproduces it; the 808-wide tile is not special.

### Expected Behavior

- `pto.tcmp` accepts a `dst` that is a packed predicate mask whose row count matches the sources and whose
  byte width covers `ceil(cols / 8)` — the same rule `pto.tcmps` already applies on a5, and the same one the
  a2a3 `pto.tcmp` path applies.
- The three `a5sim` cases reach code generation.

### Actual Behavior

- The a5 `pto.tcmp` verifier requires `src0`, `src1` and `dst` to have identical shapes, which no packed
  predicate mask can satisfy, so every tile-vs-tile compare fails to compile on a5.

### Fix

Implemented on `main` @ `cc519bc92`; 2 files, +15/-2 plus a new test.

**`lib/PTO/IR/PTO.cpp`** — in `TCmpOp::verify()`'s A5 branch, replace the declared-shape equality with the
mask-aware rule `TCmpSOp::verify()` already applies on A5 for the same mask:

```diff
-    if (getShapeVec(t0) != getShapeVec(t1) || getShapeVec(t0) != getShapeVec(td)) {
-      return emitOpError("expects src0, src1, and dst to have the same shape");
+    if (getShapeVec(t0) != getShapeVec(t1)) {
+      return emitOpError("expects src0 and src1 to have the same shape");
+    }
+
+    // dst carries a packed predicate mask, so its column extent is the packed
+    // byte width of src0's columns and does not match them. Only the row extent
+    // is shared, which is what TCmpSOp checks for the same mask.
+    auto src0Valid = getValidShapeVec(t0);
+    auto dstValid = getValidShapeVec(td);
+    if (src0Valid.size() != 2 || dstValid.size() != 2) {
+      return emitOpError("expects src0 and dst to have rank-2 valid_shape");
+    }
+    if (src0Valid[0] != ShapedType::kDynamic && dstValid[0] != ShapedType::kDynamic &&
+        src0Valid[0] != dstValid[0]) {
+      return emitOpError("expects src0 and dst to have the same valid_shape[0]");
     }
```

`src0`/`src1` shape equality is kept — both are real data tiles — and only the `dst` constraint is relaxed to
what a packed mask can satisfy.

**`test/lit/pto/compare_packed_mask_native.pto`** — new regression test with a non-byte-multiple column
count (`8x808` sources, `8x128` mask), covering `tcmp` and `tcmps` on both a5 and a3. The existing
`compare_tile_native.pto` cannot catch this: its 16-column tile and 16-byte mask happen to have equal
shapes, so it never exercises the packed path.

**Nothing else needs changing.** The lowering is shape-agnostic — `PTOCmpToEmitC` emits
`TCMP(dst, src0, src1, mode)` with the tile values and derives nothing from their shapes, exactly like
`PTOCmpSToEmitC`; `ExpandTileOp` and `InsertTemplateAttributes` only propagate `cmp_mode`. Relaxing the
verifier therefore cannot push the failure downstream, which the passing codegen below confirms.

### Test Results

Every row is the same build with only the verifier change differing.

| check | unpatched | patched |
|---|---|---|
| minimal `packed_mask.pto` (above) | `'pto.tcmp' op expects … same shape` | compiles |
| production IR `sample_gumbel_argmax.pto` | `rc=1`, no `.cpp` emitted | `rc=0`, 65358-byte `.cpp` |
| `mtp/sample.py -p a5sim` | `'pto.tcmp' op expects … same shape` | **PASS (1.40 s)** |
| `mtp/lm_head.py -p a5sim` | same error | **PASS (1.65 s)** |
| `mtp/decode_mtp.py -p a5sim` | same error | **PASS (7.24 s)** |
| new `compare_packed_mask_native.pto` | same error | lit PASS |
| existing `compare_tile_native.pto` | PASS | PASS |

The three model cases were run through pypto's normal compile path with `PTOAS_ROOT` pointed at the built
compiler, so the whole pipeline — pypto codegen, the ptoas subprocess, EmitC output — is exercised, not just
the verifier.

Not yet run: the full lit suite (1851 tests); only the five `compare_*` tests were executed.

### Environment

| Component | Version |
|---|---|
| PTOAS (failure observed) | `v0.57` (`aarch64`) |
| PTOAS (fix developed + verified) | `main` @ `cc519bc92` |
| pypto-lib CI commit | `0e7e07aa5e42db6e08c68fd5bfe1def571492b53` |
| pypto CI commit | `cb49d101aa770b72c0113f6a9ee32301f38464ff` |
| pto-isa | `39782974cea3b8068b428b43df8f0aaceb0925e4` |
| Targets | `a5sim` (fails); `a2a3sim` (passes) |
| Host platform | Linux (`aarch64`) |

### Additional Context

- **Secondary finding, not the cause: the mask dtype is not target-adjusted.** `docs/isa/TCMP.md` states the
  output type must be `uint8_t` on A2A3 but `uint32_t` on A5; pypto emits `dtype=ui8` for both. This does not
  explain the failure — a correctly-typed `u32` mask for 808 columns is `[8, 32]`, which is still not
  `[8, 808]`, so the shape check rejects it too — but it is worth confirming against the a5 kernel, which
  casts the destination to `uint32_t*` and derives its row stride as
  `DstTile::RowStride * sizeof(TOUT) / sizeof(uint32_t)`. If that division is meant to be exact, a `ui8` mask
  is only safe while the row stride is a multiple of 4. Worth a separate pypto issue if confirmed.
- The failure is at MLIR verification, before any pto-isa header is compiled, so it is independent of the
  pto-isa version — confirmed by reproducing on `master` `39782974c` as well as the CI-pinned commit.
- `sample_gumbel_argmax` is the `pl.spmd(...)` scope at `sample.py:483`; the failing compare reaches it by
  inlining `_greedy_sample_logits`.

**Comment by Crystal-wzy** [comment]
[A5_TCMP_PACKED_MASK.patch](https://github.com/user-attachments/files/31773933/A5_TCMP_PACKED_MASK.patch)
